### ISSUE TITLE
algorithm comparison

### ISSUE THREAD

**Issue by daef** [w8r/martinez#123]
not really an issue, but I wanted to dump a reference here too:

I made a small comparison between some boolean operation algorithms and
wanted to share the results: https://daef.github.io/poly-bool-comparison/

![preview](https://user-images.githubusercontent.com/2098598/77068814-83557880-69e7-11ea-9eef-b02b5997aaa3.png)

**Comment by w8r** [comment]
Great thing! can you please explain the metrics? I didn't understand the abbrevations

**Comment by daef** [comment]
I'd like to forward you to the [graphicsmagick documentation](http://www.graphicsmagick.org/compare.html) but it's not very helpful.

Wikipedia got it partially covered thou: [MAE](https://en.wikipedia.org/wiki/Mean_absolute_error), [RMSE](https://en.wikipedia.org/wiki/Root-mean-square_deviation), [PSNR](https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio)

**Comment by w8r** [comment]
Thanks a lot!

**Comment by Llorx** [comment]
Really really good tool. Thank you.

**Comment by daef** [comment]
have you found a new use case for this hack @Llorx ? i thought it doesn't do much on its own so there's no real benefit to make it more reusable.

**Comment by Llorx** [comment]
@daef Just to see the problems some libraries have with certain polygons. The red/green error deviation thingy was really useful. I was looking for a library with good rectangle handling.

**Comment by daef** [comment]
i guess when there are 'just' axis aligned rects one could do much better than with a generic polygon lib, is performance your main concern or are you looking for something correct/robust? :)

**Comment by Llorx** [comment]
Both, correctness and performant. I need to merge about 200 hundred rectangles into a single polygon in less than 16ms (for drawing to a canvas at 60 fps) but I also need for it to be drawn correctly. I thought that this new algorithm will help, but nope. I lose a lot of frames, so forgot about this solution and now I'm looking for webgl.

**Comment by w8r** [comment]
Was it too slow or incorrect?

**Comment by w8r** [comment]
Seems like rasterizing is a more convenient approach in your case.

**Comment by Llorx** [comment]
Yeah, maybe I could draw to a canvas and extract the result image as a mask. That's the next step if webgl gets too complicated haha.

In this case it was slow and incorrect, but the slowness I think is because I'm asking too much, and worse if looking for a JavaScript script instead of native code. About the correctness, I was getting some weird edges. This is the black mask image where you can see all the rectangles/squares, some of them overlap:
![image](https://user-images.githubusercontent.com/702490/85124003-7edc6e00-b229-11ea-832d-c99412e09313.png)

When the algo was passed with union, I was getting something like this (drawn with paint, but you get the point):
![image](https://user-images.githubusercontent.com/702490/85124142-b519ed80-b229-11ea-81f4-795521a8288a.png)

By the way, I noticed that is impossible for a normal computer in the browser to have a decent union performance with that amount of rectangles, so I forgot about it.

**Comment by w8r** [comment]
Yes, with "cascading" approach it's very heavy. I have faced that problem before in https://github.com/w8r/ploygon-offset

But if you are dealing with rectangles and they are axis-aligned as I see on the picture, it only makes sense to go for raster approach.

**Comment by daef** [comment]
I guess loading the rects into the GPU encoded as texture and writing a simple shader should do the trick, 200 rects doesnt sound too bad. Maybe give [this](http://madebyevan.com/shaders/fast-rounded-rectangle-shadows/) or [that](https://mortoray.com/2015/06/05/quickly-drawing-a-rounded-rectangle-with-a-gl-shader/) a shot :)

**Comment by Llorx** [comment]
> I guess loading the rects into the GPU encoded as texture and writing a simple shader should do the trick, 200 rects doesnt sound too bad. Maybe give [this](http://madebyevan.com/shaders/fast-rounded-rectangle-shadows/) or [that](https://mortoray.com/2015/06/05/quickly-drawing-a-rounded-rectangle-with-a-gl-shader/) a shot :)

Pretty nice, thank you!
